File: typescript/bc-components/src/id/uri.ts
import { type Cbor, createTag, toByteString, cbor as toCbor } from '@bc/dcbor';
import { TAG_URI } from '@bc/tags';
import type { Tag } from '@bc/dcbor';

import { BCComponentsError } from '../error.js';
import { decodeTaggedCbor, defaultTaggedCbor, defaultTaggedCborData } from '../cbor-ur.js';

const URI_TAG: Tag = createTag(TAG_URI, 'url');

export class URI {
    readonly #uri: string;

    constructor(uri: string) {
        try {
            new URL(uri);
            this.#uri = uri;
        } catch {
            throw BCComponentsError.invalidData('URI', 'invalid URI format');
        }
    }

    static new(uri: string): URI {
        return new URI(uri);
    }

    get value(): string {
        return this.#uri;
    }

    cborTags(): Tag[] {
        return [URI_TAG];
    }

    untaggedCbor(): Cbor {
        return toCbor(this.#uri);
    }

    taggedCbor(): Cbor {
        return defaultTaggedCbor(this);
    }

    taggedCborData(): Uint8Array {
        return defaultTaggedCborData(this);
    }

    static cborTags(): Tag[] {
        return [URI_TAG];
    }

    static fromUntaggedCbor(cbor: Cbor): URI {
        return new URI(cbor.toText());
    }

    static fromCbor(cbor: Cbor): URI {
        return decodeTaggedCbor(URI, cbor);
    }

    toString(): string {
        return this.#uri;
    }

    equals(other: unknown): boolean {
        return other instanceof URI && other.#uri === this.#uri;
    }
}

File: go/bccomponents/uri.go
package bccomponents

import (
	"fmt"
	"net/url"

	bctags "github.com/nickel-blockchaincommons/bctags-go"
	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
)

// URI is a Uniform Resource Identifier, validated against RFC 3986 via
// net/url.Parse.
type URI struct {
	uri string
}

// NewURI creates a new URI from a string, validating it via net/url.Parse.
func NewURI(uri string) (URI, error) {
	_, err := url.Parse(uri)
	if err != nil {
		return URI{}, errInvalidData("URI", "invalid URI format")
	}
	// url.Parse is very permissive; reject empty strings explicitly.
	if uri == "" {
		return URI{}, errInvalidData("URI", "empty URI")
	}
	return URI{uri: uri}, nil
}

// URIString returns the raw URI string.
func (u URI) URIString() string { return u.uri }

// String returns the URI string.
func (u URI) String() string { return u.uri }

// Equal reports whether two URIs are equal.
func (u URI) Equal(other URI) bool { return u.uri == other.uri }

// --- CBOR support ---

// URICBORTags returns the CBOR tags used for URI.
func URICBORTags() []dcbor.Tag {
	return dcbor.TagsForValues([]dcbor.TagValue{bctags.TagURI})
}

// CBORTags implements dcbor.CBORTagged.
func (u URI) CBORTags() []dcbor.Tag { return URICBORTags() }

// UntaggedCBOR implements dcbor.CBORTaggedEncodable.
func (u URI) UntaggedCBOR() dcbor.CBOR { return dcbor.MustFromAny(u.uri) }

// TaggedCBOR returns the tagged CBOR encoding of the URI.
func (u URI) TaggedCBOR() dcbor.CBOR {
	cbor, _ := dcbor.TaggedCBOR(u)
	return cbor
}

// ToCBOR implements dcbor.CBOREncodable.
func (u URI) ToCBOR() dcbor.CBOR { return u.TaggedCBOR() }

// DecodeURI decodes a URI from untagged CBOR.
func DecodeURI(cbor dcbor.CBOR) (URI, error) {
	s, err := cbor.TryIntoText()
	if err != nil {
		return URI{}, err
	}
	uri, err := NewURI(s)
	if err != nil {
		return URI{}, fmt.Errorf("bccomponents: %w", err)
	}
	return uri, nil
}

// DecodeTaggedURI decodes a URI from tagged CBOR.
func DecodeTaggedURI(cbor dcbor.CBOR) (URI, error) {
	return dcbor.DecodeTagged(cbor, URICBORTags(), DecodeURI)
}

// URIFromTaggedCBOR decodes a URI from tagged CBOR bytes.
func URIFromTaggedCBOR(data []byte) (URI, error) {
	return dcbor.DecodeTaggedData(data, URICBORTags(), DecodeURI)
}
